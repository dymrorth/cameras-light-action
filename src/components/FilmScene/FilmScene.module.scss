@import 'styles/common';

.backButton {
    justify-self: flex-start;
    margin-bottom: $space-s;
    padding: $space-xs * 0.5 $space-m;

    svg {
        margin-right: $space-xs;
    }
}
