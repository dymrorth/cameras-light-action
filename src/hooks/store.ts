import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import type { RootState, AppDispatch } from 'store'

import { AnyAction } from 'redux'
import { ThunkDispatch } from 'redux-thunk'

export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
export const useThunkDispatch = () => useDispatch<ThunkDispatch<RootState, void, AnyAction>>()