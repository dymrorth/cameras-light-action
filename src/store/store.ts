import { filmReducer as film } from './film/filmReducer'
import { coversReducer as covers } from './covers/coversReducer'

import { createStore, combineReducers, applyMiddleware } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension'
import thunk from 'redux-thunk'

const reducers = combineReducers({ film, covers })
const middleware = composeWithDevTools(applyMiddleware(thunk))

export const store = createStore(reducers, middleware)

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
