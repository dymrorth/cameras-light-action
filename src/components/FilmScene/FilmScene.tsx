import React, { useEffect } from 'react'

import { Film } from 'api/models/Film'
import styles from './FilmScene.module.scss'

import { useAppSelector, useThunkDispatch, useRestoreScroll } from 'hooks'
import { fetchFilm } from 'store/film/filmActions'

import { FilmPlaceholder, FilmDetail, UIError, Button } from 'components'

import { useParams, useHistory } from 'react-router'
import { IoReturnDownBackOutline } from 'react-icons/io5'

type paramsProps = { id: string | undefined }

const FilmScene: React.FC<{}> = () => {
    useRestoreScroll()

    const dispatch = useThunkDispatch()
    const film: Film = useAppSelector(state => state.film)

    const { isLoading, hasError } = useAppSelector(state => state.ui)

    const params = useParams<paramsProps>()
    const history = useHistory()

    useEffect(() => {
        if (params?.id) {
            dispatch(fetchFilm(+params.id))
        }
    }, [dispatch, params])

    const handleBackClick: React.MouseEventHandler<HTMLButtonElement> = () => {
        history.goBack()
    }

    if (isLoading) {
        return <FilmPlaceholder />
    } else {
        return (
            <UIError error={hasError}>
                <Button onClick={handleBackClick} className={styles.backButton}><IoReturnDownBackOutline /> Back</Button>
                <FilmDetail {...film}/>
            </UIError>
        )
    }
}

export default FilmScene
