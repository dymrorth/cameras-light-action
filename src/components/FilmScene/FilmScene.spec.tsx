import React from 'react'

import FilmScene from './FilmScene'
import FilmFixture from 'api/fixtures/FilmFixtures'

import { store } from 'store'

import { render, RenderResult, screen } from '@testing-library/react'
import { Provider } from 'react-redux'
import { MemoryRouter, Route } from 'react-router-dom'
import { createMemoryHistory } from 'history'

window.scroll = jest.fn()

const customRender = (ui: React.ReactElement, { route = '/', history = createMemoryHistory({ initialEntries: [route] }) }: any = {}) => {
    const Wrapper: React.FC = ({ children }) => (
        <Provider store={store}>
            <MemoryRouter initialEntries={history.initialEntries}>{children}</MemoryRouter>
        </Provider>
    )
    return {
        ...render(ui, { wrapper: Wrapper }),
        history,
    }
}

describe('<FilmScene />', () => {
    describe('when initialized', () => {
        it('has a loading FilmPlaceholder', () => {
            store.dispatch({ type: '@ui/loading' })

            const { asFragment }: RenderResult = customRender(
                <Route path="/">
                    <FilmScene />
                </Route>,
                { route: '/1' }
            )
            const loader = screen.getByTestId('FilmPlaceholder')
            expect(loader).toBeInTheDocument()
            expect(asFragment()).toMatchSnapshot()
        })
    })

    describe('when film loaded', () => {
        it('has film information', () => {
            store.dispatch({ type: '@film/fetched', film: FilmFixture.filmFromApi() })
            store.dispatch({ type: '@ui/loaded' })

            const { asFragment }: RenderResult = customRender(
                <Route path="/">
                    <FilmScene />
                </Route>,
                { route: '/1' }
            )
            const title = screen.getByText(FilmFixture.film().title)
            expect(title).toBeInTheDocument()
            expect(asFragment()).toMatchSnapshot()

        })
    })
})
