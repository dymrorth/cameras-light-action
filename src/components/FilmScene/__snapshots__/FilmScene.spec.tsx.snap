// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilmScene /> when film loaded has film information 1`] = `
<DocumentFragment>
  <button
    class="button backButton"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M112 352l-64-64 64-64"
        fill="none"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="32"
      />
      <path
        d="M64 288h294c58.76 0 106-49.33 106-108v-20"
        fill="none"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="32"
      />
    </svg>
     Back
  </button>
  <article
    class="film"
  >
    <div
      class="poster"
      style="background-image: url(undefined);"
    />
    <div
      class="main"
    >
      <div
        class="top"
      >
        <h1
          class="title"
        >
          Shang-Chi and the Legend of the Ten Rings
        </h1>
        <div>
          <div
            class="gauge rating"
            data-quarter="0"
          >
            <span
              class="value"
            >
              0
            </span>
          </div>
        </div>
      </div>
      <p
        class="overview"
      >
        Shang-Chi must confront the past he thought he left behind when he is drawn into the web of the mysterious Ten Rings organization.
      </p>
    </div>
    <div
      class="break"
    />
    <section
      class="direction"
    >
      <h3>
        Director
      </h3>
      <p
        class="namesList"
      />
    </section>
    <section>
      <h3>
        Cast
      </h3>
      <p
        class="namesList"
      />
    </section>
  </article>
</DocumentFragment>
`;

exports[`<FilmScene /> when initialized has a loading FilmPlaceholder 1`] = `
<DocumentFragment>
  <div
    class="film"
    data-testid="FilmPlaceholder"
  >
    <div
      class="poster"
    />
    <div
      class="content"
    >
      <div
        class="top"
      >
        <div
          class="title"
        />
        <div
          class="rating"
        />
      </div>
      <div
        class="overview"
      >
        <span />
        <span />
        <span />
        <span />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
